pkg/operator/ceph/cluster/mon/health.go

```diff
@@ -444,3 +498,126 @@ func (c *Cluster) findInvalidMonitorPlacement(desiredMonCount int) (*NodeUsage,
 
 	return nil, nil
 }
+
+func (c *Cluster) handleExternalMonStatus(status client.MonStatusResponse) error {
+
+	changed, err := addOrRemoveExternalMonitor(c, status)
+	if err != nil {
+		return fmt.Errorf("failed to add or remove external mon. %+v", err)
+	}
+
+	// let's save the monitor's config if anything happened
+	if changed {
+		if err := c.saveMonConfig(); err != nil {
+			return fmt.Errorf("failed to save mon config after adding/removing external mon. %+v", err)
+		}
+	}
+
+	return nil
+}
+
+func addOrRemoveExternalMonitor(c *Cluster, status client.MonStatusResponse) (bool, error) {
+	var changed bool
+	// Populate a map with the current monitor from ClusterInfo
+	monsNotFound := map[string]interface{}{}
+	for _, mon := range c.ClusterInfo.Monitors {
+		monsNotFound[mon.Name] = struct{}{}
+	}
+
+	// Iterate over the mons first and compare it with ClusterInfo
+	for _, mon := range status.MonMap.Mons {
+		inQuorum := monInQuorum(mon, status.Quorum)
+		// if the mon is not in clusterInfo
+		if _, ok := monsNotFound[mon.Name]; !ok {
+			// If the mon is part of the quorum
+			if inQuorum {
+				// let's add it to ClusterInfo
+				logger.Infof("new external mon %s found, adding it", mon.Name)
+				// find IP and Port of that Mon
+				endpoint := mon.PublicAddrs.Addrvec[0].Addr
+				monIP := cephutil.GetIPFromEndpoint(endpoint)
+				monPort := cephutil.GetPortFromEndpoint(endpoint)
+				c.ClusterInfo.Monitors[mon.Name] = cephconfig.NewMonInfo(mon.Name, monIP, monPort)
+				changed = true
+			}
+			logger.Debugf("mon %s is not in quorum and not in ClusterInfo", mon.Name)
+		} else {
+			// mon is in ClusterInfo
+			logger.Debugf("mon %s is in ClusterInfo, let's test if it's in quorum", mon.Name)
+			if !inQuorum {
+				// this mon was in clusterInfo but not part of the quorum anymore
+				// let's remove it from ClusterInfo
+				logger.Infof("monitor %s is not part of the external cluster monitor quorum, removing it", mon.Name)
```

Comment:
Does this remove the mon if it's just temporarily down? If it's still in the monmap I would want to keep it since it might just be down temporarily.

Reply:
The problem is that we can't tell if it's temporarily down or not. The check runs every 45sec so if this was temporarily it'll get re-added. Wdyt?

Reply:
What if instead of detecting the mons added and removed, could we just copy the results of the mon query to replace the local mon info? Why do we need to add and remove mons from the local list? This method would be much simpler if it was a simple translation.